server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -585,14 +587,13 @@ private long relativeTime() {
     }
 
     void processBulkIndexIngestRequest(Task task, BulkRequest original, ActionListener<BulkResponse> listener) {
-        long ingestStartTimeInNanos = System.nanoTime();
-        BulkRequestModifier bulkRequestModifier = new BulkRequestModifier(original);
-        ingestService.executeBulkRequest(() -> bulkRequestModifier,
-            (indexRequest, exception) -> {
-                logger.debug(() -> new ParameterizedMessage("failed to execute pipeline [{}] for document [{}/{}/{}]",
-                    indexRequest.getPipeline(), indexRequest.index(), indexRequest.type(), indexRequest.id()), exception);
-                bulkRequestModifier.markCurrentItemAsFailed(exception);
-            }, (exception) -> {
+        final long ingestStartTimeInNanos = System.nanoTime();
+        final BulkRequestModifier bulkRequestModifier = new BulkRequestModifier(original);
+        ingestService.executeBulkRequest(
+            original.numberOfActions(),
+            () -> bulkRequestModifier,
+            bulkRequestModifier::markItemAsFailed,
```

Comment:
I think we need to ensure that this completion handler is executed from the Write threadpool.  If I understand correctly as-is, assuming one or more async processor's are executed, this handler will execute on the last async processors' thread. We also shoud to be careful to not try to add work _from_ the write thread pool _to_ the write thread pool. 

You can see this if you make 2 processors async (for example set and uppercase) , the last async processor's thread that is executed is the one that is that is performing the BulkOperation

Reply:
Agreed, the call to `doExecute(...)` (inside completion handler) should be executed on a write thread in the case the current thread isn't a write thread.